test/rulesets/Base/fastmath_able.jl

```diff
@@ -159,28 +159,71 @@ const FASTABLE_AST = quote
-                x = -rand(T) .- 3
-                y = 3
-                Δx = randn(T)
-                Δy = randn(T)
-                Δz = randn(T)
-
-                @test frule((ZeroTangent(), Δx, Δy), ^, x, y)[2] ≈ Δx * y * x^(y - 1)
-                @test frule((ZeroTangent(), Δx, Δy), ^, zero(x), y)[2] ≈ 0
-                _, ∂x, ∂y = rrule(^, x, y)[2](Δz)
-                @test ∂x ≈ Δz * y * x^(y - 1)
-                @test ∂y ≈ 0
-                _, ∂x, ∂y = rrule(^, zero(x), y)[2](Δz)
-                @test ∂x ≈ 0
-                @test ∂y ≈ 0
+        @testset "^(x::$T, p::$S)" for T in (Float64, ComplexF64), S in (Float64, ComplexF64)
+            test_frule(^, rand(T) + 3, rand(S) + 3)
+            test_rrule(^, rand(T) + 3, rand(S) + 3)
+
+            # When both x & p are Real, and !(isinteger(p)), 
+            # then x must be positive to avoid a DomainError
+            T <: Real && S <: Real && continue
+            # In other cases, we can test values near zero:
+
+            test_frule(^, randn(T), rand(S))
+            test_rrule(^, rand(T), rand(S))
+        end
+
+        # Tests for power functions, at values near to zero.
+        POWERGRADS = [ # (x,p) => (dx,dp)
+        # Some regular points, as sanity checks:
+          (1.0, 2)   => (2.0, 0.0),
+          (2.0, 2)   => (4.0, 2.772588722239781),
+        # At x=0, gradients for x seem clear, 
+        # for p less certain what's best.
+          (0.0, 2)   => (0.0, 0.0),
+          (-0.0, 2)  => (0.0, 0.0),  # probably (-0.0, 0.0) would be ideal
+          (0.0, 1)   => (1.0, 0.0),
+          (-0.0, 1)  => (1.0, 0.0),
+          (0.0, 0)   => (0.0, NaN),
+          (-0.0, 0)  => (0.0, NaN),
+          (0.0, -1)  => (-Inf, NaN),
+          (-0.0, -1) => (-Inf, NaN),
+          (0.0, -2)  => (-Inf, NaN),
+          (-0.0, -2) => (Inf, NaN),
+        # Integer x & p, check no InexactErrors
+          (0, 2)    => (0.0, 0.0),
+          (0, 1)    => (1.0, 0.0),
+          (0, 0)    => (0.0, NaN),
+          (0, -1)   => (-Inf, NaN),
+          (0, -2)   => (-Inf, NaN),
+        # Non-integer powers:
+          (0.0, 0.5)   => (Inf, 0.0),
+          (0.0, 3.5)   => (0.0, 0.0),
+          (0.0, -1.5)  => (-Inf, NaN),
+        ]
+
+        @testset "$x ^ $p" for ((x,p), (∂x, ∂p)) in POWERGRADS
+            if x isa Integer && p isa Integer && p < 0
+                @test_throws DomainError x^p
+                continue
             end
+            y = x^p
+
+            # Forward
+            y_fwd = frule((1,1,1), ^, x, p)[1]
+            @test isequal(y, y_fwd)
+
+            # ∂x_fwd = frule((0,1,0), ^, x, p)[1]
+            # ∂p_fwd = frule((0,0,1), ^, x, p)[2]
+            # isequal(∂x, ∂x_fwd) || println("^ forward `x` gradient for $y = $x^$p: got $∂x_fwd, expected $∂x, maybe!")
+            # isequal(∂p, ∂p_fwd) || println("^ forward `p` gradient for $x^$p: got $∂p_fwd, expected $∂p, maybe")
```

Comment:
It's marked draft because these frule tests aren't quite right.

Reply:
Ah, I see!

Reply:
The print statements give this:
```
^ forward `x` gradient for 1.0 = 1.0^2: got 1.0, expected 2.0, maybe!
^ forward `x` gradient for 0.0 = 0.0^1: got 0.0, expected 1.0, maybe!
^ forward `x` gradient for -0.0 = -0.0^1: got -0.0, expected 1.0, maybe!
^ forward `x` gradient for 1.0 = 0.0^0: got 1.0, expected 0.0, maybe!
^ forward `x` gradient for 1.0 = -0.0^0: got 1.0, expected 0.0, maybe!
^ forward `x` gradient for Inf = 0.0^-1: got Inf, expected -Inf, maybe!
^ forward `x` gradient for Inf = 0.0^-2: got Inf, expected -Inf, maybe!
^ forward `x` gradient for 0 = 0^1: got 0, expected 1.0, maybe!
^ forward `x` gradient for 1 = 0^0: got 1, expected 0.0, maybe!
^ forward `x` gradient for 0.0 = 0.0^0.5: got 0.0, expected Inf, maybe!
^ forward `p` gradient for 0.0^0.5: got NaN, expected 0.0, maybe
^ forward `x` gradient for Inf = 0.0^-1.5: got Inf, expected -Inf, maybe!
```

Reply:
Maybe fixed, by regarding input `Δp==0` as a "strong zero". I'm not 100% sure this is idea but someone else can fix it if they desire.

Reply:
The diffractor issue is forward mode. My version:
```
function f1(x) # from https://github.com/JuliaDiff/Diffractor.jl/issues/26
    res = 0.0
    for i in 1:10
        res += x^i
    end
    res
end

using Diffractor, ForwardDiff, Plots
f1fwd(x) = let var"'" = Diffractor.PrimeDerivativeFwd 
    f1'(x)
end
f1rev(x) = let var"'" = Diffractor.PrimeDerivativeBack
    f1'(x)
end
xs = -0.5:0.01:0.5
p1 = plot(xs, f1.(xs), lab="f1")
plot!(xs, f1fwd.(xs), lab="f1', forward")
plot!(xs, ForwardDiff.derivative.(f1, xs).+0.02, lab="f1', dual")
plot!(xs, f1rev.(xs).+0.04, lab="f1', reverse")  # fails with this PR
```
Before:
![f1_dual](https://user-images.githubusercontent.com/32575566/132075909-01b60b47-efeb-45ab-9613-e71d4d9b8338.png)

After:
![f1_after](https://user-images.githubusercontent.com/32575566/132075952-8c2b1fb3-f445-4e55-851e-5ba1b5aa51ec.png)

So the original problem is fixed. But there was also a worse problem with reverse mode. Which now doesn't run at all after this PR.

Reply:
Oh now I'm less sure. That's on Julia 1.8-, mac M1 native. On 1.7, rosetta, I can't plot, but I don't see the problem. With or without this PR, forward and reverse agree:
```
julia> f1fwd(0.5)
3.9765625

julia> f1rev(0.5)
3.9765625
```